### PR TITLE
Support session reuse in RedshiftDataOperator

### PR DESCRIPTION
**Pull request by borismo**
Recently, AWS [added](https://aws.amazon.com/about-aws/whats-new/2024/09/session-reuse-amazon-redshift-data-api/) [session reuse](https://docs.aws.amazon.com/redshift/latest/mgmt/data-api.html#data-api-calling-considerations-session-reuse) to the Redshift Data API. It allows to, say, create a temporary table in one statement, and select from it in a subsequent one.
I think it would be useful that `RedshiftDataOperator` supports this new feature.

# Decisions
- the hook and operator's `database` arguments now optional because when the session ID is provided, boto3 doesn't allow database to be specified. Because this becomes a named argument, I had to move it after `sql`. If users of the operator and hook rely on the position to pass the argument that could break things ⚠️ . Note that [boto3 only requires Sql](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/redshift-data/client/execute_statement.html).
- the hook now returns an object containing the statement ID and session ID instead of just a statement ID. I choose a `dataclass` instead of a dict to make it harder to make a mistake when indexing the keys. If users use the hook in their tasks or custom operators, this could also break things ⚠️ 
- fixed the typo in the `parse_statement_resposne` method. It's a public one. So, again, there is a risk of breaking change. ⚠️ 
- boto3's validation error messages are not very useful when the session ID is a non-UUID string, or if no database, workgroup or session ID was provided, so I added checks in the hook.

# To do
- [x] update docs
- [x] update changelog
- [x] update and run examples
  - [x] example_redshift.py
  - [x] example_redshift_s3_transfers.py

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/operators/redshift_data.py [hunk 1]
```diff
@@ -97,6 +98,8 @@ def __init__(
         return_sql_result: bool = False,
         workgroup_name: str | None = None,
         deferrable: bool = conf.getboolean("operators", "default_deferrable", fallback=False),
+        session_id: str | None = None,
+        session_keep_alive_seconds: int | None = None,
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/redshift_data.py:105]: Please add them to docstring

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 2]
```diff
@@ -76,7 +88,9 @@ def execute_query(
         wait_for_completion: bool = True,
         poll_interval: int = 10,
         workgroup_name: str | None = None,
-    ) -> str:
+        session_id: str | None = None,
+        session_keep_alive_seconds: int | None = None,
+    ) -> QueryExecutionOutput:
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:93]: This is a breaking change. Users who are using `RedshiftDataHook.execute_query` will have their DAG fail with this change. I dont have a solution though (besides creating a new operator which I really dont like). Or maybe creating a new major version is fine? It has been a while we have not cut a new major release of the Amazon provider package. @o-nikolas @ferruzzi ?

**Review by vincbeck — COMMENTED** [review]
Apart from your new feature, you seem to fix many bugs, thanks for doing it!

File: airflow/providers/amazon/aws/hooks/redshift_data.py (hunk 2 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/hooks/redshift_data.py:93]: Yeah, I wouldn't be opposed to a major version bump, it has been over a year without a major version bump. Do we have a wishlist of other things we'd like to change as well in this bump? It's an opportunity to do some refactoring and drop some deprecated code.

Comment (vincbeck) [reply airflow/providers/amazon/aws/hooks/redshift_data.py:93]: Agree, I dont think there is but we should take advantage of this opportunity

**Review by vincbeck — CHANGES_REQUESTED** [review]
Setting "request changes" so that it is not merged out of the blue. Since this PR contains a breaking change, that means the next release of the amazon provider package will be a major one. We (AWS team) want to take this opportunity to remove deprecated stuff. This is just a way to control when this PR will be merged

**Comment by vincbeck** [comment]
@borismo Just checking in, are you working on this one?

**Comment by borismo** [comment]
@vincbeck , yes, but on my free time. I'll try to make some progress soon.

**Comment by borismo** [comment]
The `Compat 2.` checks fail if the test expects `execution_date=None` to be pushed to the Xcom.
`Non-DB::3.8` and `All:LowestDeps-Postgres:12:3.8` fail when I add it.
Does anyone know how to solve this?

**Comment by vincbeck** [comment]
> The `Compat 2.` checks fail if the test expects `execution_date=None` to be pushed to the Xcom. `Non-DB::3.8` and `All:LowestDeps-Postgres:12:3.8` fail when I add it. Does anyone know how to solve this?

Interesting, I'd like to see the error from compat tests. Could you solve the non-db tests please?

**Comment by borismo** [comment]
> Could you solve the non-db tests please?

Sure, done in https://github.com/apache/airflow/pull/42218/commits/87da7b95cef97998887c1f57fc230ff2fc930e5d

File: tests/providers/amazon/aws/operators/test_redshift_data.py [hunk 3]
```diff
@@ -163,6 +173,58 @@ def test_execute_with_workgroup_name(self, mock_exec_query):
...
+            with_event=False,
+            wait_for_completion=wait_for_completion,
+            poll_interval=poll_interval,
+            session_id=None,
+            session_keep_alive_seconds=123,
+        )
+        mock_ti.xcom_push.assert_called_once_with(
+            key="session_id",
+            value=SESSION_ID,
         )
```
Comment (vincbeck) [inline tests/providers/amazon/aws/operators/test_redshift_data.py]: This one should work in non-db tests and compat tests

```suggestion
        assert mock_ti.xcom_push.call_args.kwargs["key"] == "session_id"
        assert mock_ti.xcom_push.call_args.kwargs["value"] == SESSION_ID
```

**Comment by vincbeck** [comment]
That worked :) Merging it now

**Review by vincbeck — APPROVED** [review]

**Comment by vincbeck** [comment]
@eladkal This is a breaking change in Amazon provider package. So you know, the next release should be a major version bump. AWS team will take advantage of this to remove deprecated stuff from Amazon provider package

**Pull request merged**